### ISSUE TITLE
[Feature Request]: 

### ISSUE THREAD

**Issue by shichunshan** [infiniflow/ragflow#1268]
### Is there an existing issue for the same feature request?

- [X] I have checked the existing issues.

### Is your feature request related to a problem?

```Markdown
./entrypoint.sh: line 25: 18337 Killed                  $PY api/ragflow_server.py
[WARNING] [2024-06-24 07:59:24,499] [synonym.__init__] [line:24]: Realtime synonym is disabled, since no redis connection.
SUM: [WARNING] Load term.freq FAIL!

    ____                 ______ __               
   / __ \ ____ _ ____ _ / ____// /____  _      __
  / /_/ // __ `// __ `// /_   / // __ \| | /| / /
 / _, _// /_/ // /_/ // __/  / // /_/ /| |/ |/ / 
/_/ |_| \__,_/ \__, //_/    /_/ \____/ |__/|__/  
              /____/                             

    
[INFO] [2024-06-24 07:59:36,021] [_internal._log] [line:96]: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
```

### Describe the feature you'd like

run_simple causes the service to be killed, interrupting the service. I hope to be able to deploy using WSGI.

**Comment by r1ng13** [comment]
想问下这个问题解决了吗

**Comment by bwnjnOEI** [comment]
+1

**Comment by a1299646329** [comment]
+1

**Comment by tuoxiebushuijiao** [comment]
+1

**Comment by lichanghai** [comment]
+1

**Comment by pan003** [comment]
mark

**Comment by zanyatta** [comment]
+1

**Comment by zyoung1212** [comment]
+1

**Comment by ericwu0930** [comment]
I’m unsure if this is directly related, but the concurrency performance is poor.

When making a single call, the response time is around 1 second. However, when I initiated 100 concurrent requests, the performance degraded significantly (see attached results).
![Image](https://github.com/user-attachments/assets/00c24590-9314-4e12-a3a4-ebd65639fafd)

The code:
```python
async def test_get_fund_manager():
    import asyncio
    
    async def single_query(i):
        start_time = time.time()
        result = await Rag_Api.retrieval_rag("请问如何撤单")
        end_time = time.time()
        duration = end_time - start_time
        
        logger.info(f"第{i+1}次查询耗时: {duration:.2f}秒")
        return result, duration

    tasks = [single_query(i) for i in range(100)]
    results = await asyncio.gather(*tasks)
    
    total_duration = sum(duration for _, duration in results)
    logger.info(f"并发查询总耗时: {total_duration:.2f}秒")


```
@which-W

**Comment by ZhenhangTung** [comment]
Related PR: https://github.com/infiniflow/ragflow/pull/7845  And thanks @ericwu0930 for the details. Any further thoughts? @zyoung1212 @zanyatta @pan003 @lichanghai @tuoxiebushuijiao @a1299646329 @bwnjnOEI @r1ng13 @shichunshan :)

**Comment by ericwu0930** [comment]
@ZhenhangTung Coincidentally, I also tried using gunicorn instead of the original wsgi server and saw a 10x increase in concurrency performance. I hope to release this feature soon.

**Comment by ZhenhangTung** [comment]
> [@ZhenhangTung](https://github.com/ZhenhangTung) Coincidentally, I also tried using gunicorn instead of the original wsgi server and saw a 10x increase in concurrency performance. I hope to release this feature soon.

@ericwu0930 I saw @zyoung1212 has just finished work. You can leave your thoughts there. Could you also explain how you achieve 10x increase in concurrency performance? Any benchmark details?

**Comment by ericwu0930** [comment]
> > [@ZhenhangTung](https://github.com/ZhenhangTung) Coincidentally, I also tried using gunicorn instead of the original wsgi server and saw a 10x increase in concurrency performance. I hope to release this feature soon.
> 
> [@ericwu0930](https://github.com/ericwu0930) I saw [@zyoung1212](https://github.com/zyoung1212) has just finished work. You can leave your thoughts there. Could you also explain how you achieve 10x increase in concurrency performance? Any benchmark details? 

- Concurrency: 30
- Duration: 1 minute
- API endpoint: /api/v1/retrieval
- cpu: Intel Xeon Glod 6134@3.20GHz 16 cores
- ram: 64GB
- same question

default wsgi server:
![Image](https://github.com/user-attachments/assets/871c52d1-9da3-4cb2-8fa4-b2572b7bf252)
use gunicorn --workers=20 to start ragflow server:
![Image](https://github.com/user-attachments/assets/f473a32d-2b5d-49ed-ae52-ec4ff741f949)

What's more, I use the default rerank algorithm rather than set a rerank model. I suspect the poor concurrency performance stems from:
1. The default wsgi server(werkzeug) running in thread mode;
2. The default rerank algorithm is highly time-consuming and being CPU-bound (I have mentioned at #7805, #7746
but no one reply me lolll);
3. Task blocking due to Python’s GIL and poor performance of werkzeug.

**Comment by KevinHuSh** [comment]
@ericwu0930 
The median of retrieval time for a query is 6009 ms isn't it?
Did you use default embedding model which was deployed on CPU or other embedding SAAS?

**Comment by ericwu0930** [comment]
@KevinHuSh 
Yes, when I use the default wsgi server, the average is 50s. After the modification, the median is 6s.
I think switching the wsgi server to gunicorn could solve RAGFlow's concurrency issues, but are there ways to improve single retrieval performance?
I used an embedding SaaS service from aliyun, and from what I observed, the embedding process doesn't consume much time.
By adding logs, I concluded the following time consumption:
In a typical 5-second retrieval operation:
Embedding the question takes 300ms
Fetching answers from ES takes 200ms
The remaining 4+ seconds are spent on re-ranking the retrieved results.

**Comment by KevinHuSh** [comment]
@ericwu0930 
Did you utilize rerank model?

**Comment by ericwu0930** [comment]
@KevinHuSh 
No, I didn't set the rerank model.  I think it uses the default rerank algorithm.

![Image](https://github.com/user-attachments/assets/362fcc3b-c2e6-41bb-903d-19f63dc77688)

**Comment by yingfeng** [comment]
Will use golang for better performance soon
